integration-cli/docker_cli_build_test.go

```diff
@@ -5660,3 +5660,434 @@ func (s *DockerSuite) TestBuildNullStringInAddCopyVolume(c *check.C) {
 	_, err = buildImageFromContext(name, ctx, true)
 	c.Assert(err, check.IsNil)
 }
+
+func (s *DockerSuite) TestBuildBuildTimeArg(c *check.C) {
+	imgName := "bldargtest"
+	envKey := "foo"
+	envVal := "bar"
+	args := []string{
+		"--build-arg", fmt.Sprintf("%s=%s", envKey, envVal),
+	}
+	dockerfile := fmt.Sprintf(`FROM busybox
+		ARG %s=
```

Comment:
@tiborvass 

> although allowing an empty default value for ARG foo would make some sense, it could bite us in the future if, for some reason, we ever want to allow "no default values".

I agree and when I was referring to allowing the `ARG foo` syntax, I meant that it defines the ARG but *doesn't* assigns a default value. There is a subtle distinction between defining and not assigning a default v/s defining and always assigning a default (https://github.com/docker/docker/pull/15182#discussion_r38440476) that I think you are also referring to. 

And I believe we need to support `ARG`s that are defined but not have a default value for this feature to work *irrespective of the syntax* that we agree on. It is needed to provide a consistent behavior with pre-defined ARGs that are defined but don't have a default value (like proxy, which if I am correct will break the build if we inject empty values for them).

If this makes sense then, as we have figured so far from the discussions, following appear to be the options to achieve `ARG`s that are defined with no default value:
- only allow `ARG key=` syntax (as provided by current patch)
- allow `ARG key` as well as `ARG key=` syntax to define `ARG`s but not assign a default value
- allow `ARG key` to define but not assign a default value and then treat `ARG=` same as `ARG=""` i.e. define and set default value to `""`

I think @duglin is inclining towards the last option. I am fine with that as well. WDYT? 

Sorry am not implying to come out aggressive by listing the options but just trying keep the context around as we progress and eliminate/keep certain options.